# Load all the necessary libraries, these should already be installed.

### If you need additional libraries, add them to the dependencies.r files and rebuild the image


In [1]:
library(knitr)
library(plotly)
library(httr)
library(jsonlite)
library(purrr)
library(magrittr)
library(GetoptLong)

Loading required package: ggplot2

Attaching package: ‘plotly’

The following object is masked from ‘package:ggplot2’:

    last_plot

The following object is masked from ‘package:stats’:

    filter

The following object is masked from ‘package:graphics’:

    layout


Attaching package: ‘httr’

The following object is masked from ‘package:plotly’:

    config


Attaching package: ‘purrr’

The following object is masked from ‘package:jsonlite’:

    flatten


Attaching package: ‘magrittr’

The following object is masked from ‘package:purrr’:

    set_names



# Functions that will eventually be in an R library and maintained alongside the API

### This is PIC-SURE-HPDS specific code, the same functionality is available through PIC-SURE 2's HPDS resource, but this saves you the PIC-SURE ceremonial wrapping of JSON in more JSON.


In [2]:

escapeSlashes <- function(field){
    return (gsub("\\", "\\\\", field, fixed=TRUE))
}

wrapInQuotes <- function(value){
    return (paste('"', escapeSlashes(value), '"', sep=''))
}

numericFilter <- function(field, min, max){
    field <- escapeSlashes(field)
    return (qq('"@{field}":{"min":@{min},"max":@{max}}'))
}

categoryFilter <- function(field, categories){
    field <- escapeSlashes(field)
    categories <- paste(map(categories, wrapInQuotes), collapse=', ')
    return (qq('"@{field}":[@{categories}]'))
}


dictionary <- function(query){
    response <- POST("http://pic-sure-hpds-nhanes:8080/PIC-SURE/search/", body=paste('{"query":"',query,'"}', sep=""), content_type_json())
    response
    mergeCategoryValues <- function(entry){
        entry$categoryValues
        entry$categoryValues <- paste(entry$categoryValues, collapse=", ")
        return (entry)        
    }
    
    return(map_df(content(response)$results, mergeCategoryValues))
}    

dictionaryFields <- function(query){
    as.list(dictionary(query)$name)
}

query <- function(selectedFields = NULL, requiredFields = NULL, numericFilters = NULL, categoryFilters = NULL) {
    fields <- function(selectedFields){
        paste(map(selectedFields, wrapInQuotes), collapse=', ')
    }

    query <- qq(
        '{"fields":[@{fields}],
        "requiredFields":[@{requiredFields}],
        "numericFilters":{@{numericFilters}},
        "categoryFilters":{@{categoryFilters}}}', 
          envir = list(
              fields = fields(selectedFields), 
              requiredFields = fields(requiredFields), 
              numericFilters = paste(numericFilters, sep=","), 
              categoryFilters = paste(categoryFilters, sep=",")))
    query
    response <- POST("http://pic-sure-hpds-nhanes:8080/query", body=query, content_type_json())
    content(response)
}

getStatus <- function(queryId){
    response <- GET(paste("http://pic-sure-hpds-nhanes:8080/query", queryId, "status", sep="/"))
    content(response)
}

getResult <- function(queryId){
    return(read.csv(paste("http://pic-sure-hpds-nhanes:8080/query", queryId, "result", sep="/")))
}

runQueryAndGetResult <- function(selectedFields = NULL, requiredFields = NULL, numericFilters = NULL, categoryFilters = NULL){
    queryStatus <- query(selectedFields, requiredFields, numericFilters, categoryFilters)
    while(queryStatus$status == "RUNNING" || queryStatus$status == "PENDING"){
        Sys.sleep(.1)
        queryStatus <- getStatus(queryStatus$id)
    }
    if(queryStatus$status == "SUCCESS" && queryStatus$numRows > 0){
        print(paste(
            queryStatus$numRows, " rows and ", 
            queryStatus$numColumns, " columns ", 
            "completed in : ", (queryStatus$completedTime - queryStatus$queuedTime)/1000, " seconds", sep=""))
        return (getResult(queryStatus$id))
    }else{
        queryStatus
    }
}


# Let's start by looking at the whole dictionary file

## Notice that the important data is all here, if it's continuous we get min and max values, if it's categorical we get all the available categories for filtering. We also get a count of observations, at this point this is a patient count only because we haven't dealt with temporal relationships yet.

In [3]:
dictionary("demographics")

categorical observationCount      max min                        name
1        FALSE            41474      4.0   1  \\demographics\\SDDSRVYR\\
2        FALSE            38076      5.0   0  \\demographics\\INDFMPIR\\
3        FALSE            38076      2.0   0 \\demographics\\SES_LEVEL\\
4         TRUE            41474       NA  NA       \\demographics\\SEX\\
5        FALSE            41445      3.0   1   \\demographics\\DMDBORN\\
6        FALSE            21004 103831.2   0  \\demographics\\WTMEC4YR\\
7        FALSE            26793     99.0   1  \\demographics\\DMDMARTL\\
8        FALSE            41474     85.0   0       \\demographics\\AGE\\
9        FALSE            41474      3.0   1   \\demographics\\SDMVPSU\\
10        TRUE            41474       NA  NA      \\demographics\\RACE\\
11       FALSE            41474 261361.3   0  \\demographics\\WTMEC2YR\\
12       FALSE            41474     58.0   1  \\demographics\\SDMVSTRA\\
13       FALSE            34260      2.0   0 \\demographics\\education\\
14        TRUE            41474       NA  NA      \\demographics\\area\\
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              categoryValues
1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           
2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           
3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           
4                                                                                                                                                                       

# Let's look at something more interesting... all variables related to "blood"


In [4]:
dictionary("blood")

categorical observationCount       max     min
1        FALSE             4273 7.071e-01  0.0707
2        FALSE             5369 6.270e+03  0.1344
3        FALSE             4486 6.790e-02  0.0354
4        FALSE            33604 5.300e+00  0.0000
5        FALSE             4047 1.870e+01  0.0495
6        FALSE             4435 5.600e-02  0.0064
7        FALSE             5616 2.490e+02  0.0260
8        FALSE             5573 7.470e-01  0.0035
9        FALSE            33718 1.970e+01  5.8000
10       FALSE            26036 1.320e+02  0.0000
11       FALSE             5778 3.860e+00  0.0064
12       FALSE             5797 9.680e+01  0.0177
13       FALSE             4414 1.344e-01  0.0707
14       FALSE             4488 3.800e-02  0.0057
15       FALSE             5679 5.460e+00  0.0100
16       FALSE            33718 4.390e+01 14.7000
17       FALSE            33605 3.980e+01  0.6000
18       FALSE             5671 2.500e+00  0.0170
19       FALSE             4168 7.070e-02  0.0141
20       FALSE            33718 1.219e+02 51.1000
21       FALSE            33605 5.720e+01  0.0000
22       FALSE             4510 1.600e-01  0.0212
23        TRUE            18488        NA      NA
24       FALSE             5367 2.200e+04  1.4849
25       FALSE             1369 1.340e-02  0.0071
26       FALSE             5945 6.420e+00  0.0170
27       FALSE            33718 4.490e+01 27.8000
28       FALSE            33604 4.500e+00  0.0000
29       FALSE            33717 1.350e+01  5.0000
30       FALSE            33604 1.810e+01  0.2000
31       FALSE             4564 1.340e-02  0.0071
32       FALSE            33605 8.970e+01  4.5000
33       FALSE            33605 9.270e+01  2.4000
34       FALSE             3165 7.100e-03  0.0071
35       FALSE            33718 5.990e+01 16.3000
36       FALSE            33718 9.160e+00  1.7100
37       FALSE             5858 1.940e+01  0.0240
38       FALSE            33717 1.000e+03  4.0000
39       FALSE             5463 5.400e+02  0.3451
40       FALSE            33717 9.999e+01  2.3000
41       FALSE             4564 7.900e-02  0.0064
42       FALSE             4384 7.700e-01  0.0078
43       FALSE             3131 7.100e-03  0.0071
44       FALSE             4442 1.480e-02  0.0078
45       FALSE            33604 4.500e+00  0.0000
46       FALSE             4510 1.340e-02  0.0071
47       FALSE            26036 2.700e+02 72.0000
48       FALSE            33605 3.540e+01  0.0000
49       FALSE             4525 1.480e-02  0.0078
50       FALSE             5526 1.000e+02  0.0066
51       FALSE             8019 3.040e+02  0.0000
52       FALSE             5622 8.110e+01  0.1407
53       FALSE            33718 3.160e+01 10.5000
54       FALSE             6700 1.500e+01  0.0071
55       FALSE             4534 1.200e-02  0.0064
56       FALSE            26047 1.220e+02  1.0000
57       FALSE             5701 4.000e+01  0.0144
                                                                               name
1               \\laboratory\\volatile compounds\\Blood Nitrobenezene (ng per mL)\\
2                        \\laboratory\\volatile compounds\\Blood MTBE (pg per mL)\\
3         \\laboratory\\volatile compounds\\Blood 1,3-Dichlorobenzene (ng per mL)\\
4                                         \\laboratory\\blood\\Eosinophils number\\
5          \\laboratory\\volatile compounds\\Blood Methylene Chloride (ng per mL)\\
6          \\laboratory\\volatile compounds\\Blood 1,2-Dichloroethane (ng per mL)\\
7         \\laboratory\\volatile compounds\\Blood 1,4-Dichlorobenzene (ng per mL)\\
8        \\laboratory\\volatile compounds\\Blood Carbon Tetrachloride (ng per mL)\\
9                                      \\laboratory\\blood\\Hemoglobin (g per dL)\\
10                                  \\examination\\blood pressure\\mean diastolic\\
11            \\laboratory\\volatile compounds\\Blood Trichloroethene (ng per mL)\\
12                    \\laboratory\\volatile compounds\\Blood Toluene (ng per mL)\\
13        \

# OK, so how about all that data for "blood"? 

### Notice here we just pass the result of our dictionary search, but by using the convenient dictionaryFields function that handles escaping our terms for us and building a vector of fields. 

In [5]:

data = runQueryAndGetResult(dictionaryFields("blood"))

[1] "41474 rows and 58 columns completed in : 1.22 seconds"


In [6]:
data

Patient.ID X.examination.blood.pressure.60.sec.HR..30.sec.HR...2..
1              1                                                     NaN
2              2                                                     NaN
3              3                                                     126
4              4                                                     NaN
5              5                                                     NaN
6              6                                                     106
7              7                                                     NaN
8              8                                                     NaN
9              9                                                     NaN
10            10                                                     NaN
11            11                                                     NaN
12            12                                                     NaN
13            13                                                     NaN
14            14                                                     NaN
15            15                                                     NaN
16            16                                                     NaN
17            17                                                     NaN
18            18                                                     NaN
19            19                                                     NaN
20            20                                                     NaN
21            21                                                     102
22            22                                                     NaN
23            23                                                     NaN
24            24                                                     122
25            25                                                     NaN
26            26                                                     NaN
27            27                                                     NaN
28            28                                                     NaN
29            29                                                     NaN
30            30                                                     NaN
⋮              ⋮                                                       ⋮
41445      41445                                                     116
41446      41446                                                     NaN
41447      41447                                                     NaN
41448      41448                                                     NaN
41449      41449                                                     120
41450      41450                                                     NaN
41451      41451                                                     106
41452      41452                                                     NaN
41453      41453                                                     NaN
41454      41454                                                     NaN
41455      41455                                                     NaN
41456      41456                                                     NaN
41457      41457                                                     NaN
41458      41458                                                     NaN
41459      41459                                                     100
41460      41460                                                     NaN
41461      41461                                                     NaN
41462      41462                                                     NaN
41463      41463                                                     NaN
41464      41464                                                     NaN
41465      41465                                                     NaN
41466      41466                                                     NaN
41467      41467                                                

# OK, so how about all that data for "blood" but only where patients have a Basophils percent? 

### Notice we are requiring patients to have all fields that match the search "Basophils percent". If we try this for the search term "blood" it will require all fields that have blood in their path, which matches no patients so we would get no results.

In [7]:
data = runQueryAndGetResult(selectedFields=dictionaryFields("blood"), 
                     requiredFields=dictionaryFields("Basophils percent"))


[1] "33605 rows and 58 columns completed in : 0.53 seconds"


In [8]:
data

Patient.ID X.examination.blood.pressure.60.sec.HR..30.sec.HR...2..
1              1                                                     NaN
2              2                                                     NaN
3              4                                                     NaN
4              5                                                     NaN
5              6                                                     106
6              7                                                     NaN
7              8                                                     NaN
8              9                                                     NaN
9             10                                                     NaN
10            11                                                     NaN
11            12                                                     NaN
12            13                                                     NaN
13            15                                                     NaN
14            16                                                     NaN
15            17                                                     NaN
16            18                                                     NaN
17            19                                                     NaN
18            20                                                     NaN
19            21                                                     102
20            22                                                     NaN
21            23                                                     NaN
22            24                                                     122
23            25                                                     NaN
24            26                                                     NaN
25            27                                                     NaN
26            28                                                     NaN
27            29                                                     NaN
28            30                                                     NaN
29            31                                                     NaN
30            32                                                     NaN
⋮              ⋮                                                       ⋮
33576      41437                                                     NaN
33577      41438                                                     NaN
33578      41439                                                     NaN
33579      41440                                                     NaN
33580      41441                                                     NaN
33581      41442                                                     116
33582      41443                                                     NaN
33583      41444                                                     NaN
33584      41447                                                     NaN
33585      41448                                                     NaN
33586      41450                                                     NaN
33587      41451                                                     106
33588      41452                                                     NaN
33589      41453                                                     NaN
33590      41455                                                     NaN
33591      41456                                                     NaN
33592      41457                                                     NaN
33593      41458                                                     NaN
33594      41460                                                     NaN
33595      41462                                                     NaN
33596      41465                                                     NaN
33597      41466                                                     NaN
33598      41467                                                

# Right... 33605 patients, let's try the same query, but only age 5-25

In [9]:
numFilters <- c(numericFilter("\\demographics\\AGE\\",5,25))
data = runQueryAndGetResult(selectedFields=dictionaryFields("blood"), 
                     requiredFields=dictionaryFields("Basophils percent"), 
                     numericFilters=numFilters)

[1] "14542 rows and 59 columns completed in : 3.007 seconds"


In [10]:
data

Patient.ID X.demographics.AGE.
1              5                  18
2             11                  17
3             19                  12
4             22                  20
5             25                  18
6             26                  23
7             29                  17
8             30                  19
9             33                  12
10            38                  19
11            44                   5
12            45                  16
13            47                  17
14            48                  12
15            52                  11
16            53                  19
17            56                  15
18            59                  19
19            61                  17
20            62                   8
21            65                   7
22            66                  16
23            71                  11
24            73                  12
25            74                  24
26            83                  13
27            85                  16
28            87                  11
29            88                  13
30            92                  18
⋮              ⋮                   ⋮
14513      41384                  15
14514      41386                  17
14515      41396                  13
14516      41397                  18
14517      41398                   7
14518      41399                  13
14519      41400                  12
14520      41401                  12
14521      41402                   6
14522      41407                  13
14523      41410                  17
14524      41411                  22
14525      41413                  10
14526      41415                  12
14527      41417                  16
14528      41421                   6
14529      41422                  18
14530      41432                  13
14531      41436                   5
14532      41438                  24
14533      41439                   9
14534      41440                  19
14535      41444                  17
14536      41455                  15
14537      41457                  16
14538      41462                  19
14539      41469                  14
14540      41470                  18
14541      41473                  18
14542      41474                   5
      X.examination.blood.pressure.60.sec.HR..30.sec.HR...2..
1                                                         NaN
2                                                         NaN
3                                                         NaN
4                                                         NaN
5                                                         NaN
6                                                         NaN
7                                                         NaN
8                                                         NaN
9                                                         NaN
10                                                        NaN
11                                                         84
12                                                        NaN
13                                                        NaN
14                                                        NaN
15                                                        NaN
16                                                        NaN
17                                                        NaN
18                                                        NaN
19                                                        NaN
20                                                        NaN
21                                                         84
22                                                        NaN
23                                                        NaN
24                                                        NaN
25                                                        NaN
26                                                        NaN
27                                    

# This is better, what about only Male patients who are of white or mexican decent and we want to add all lab values?

In [11]:
numFilters <- c(numericFilter("\\demographics\\AGE\\",5,25))
catFilters <- c(categoryFilter("\\demographics\\RACE\\", c("white", "mexican")))
data = runQueryAndGetResult(selectedFields=union(dictionaryFields("laboratory"), dictionaryFields("blood")), 
                     requiredFields=dictionaryFields("Basophils percent"), 
                     numericFilters=numFilters, categoryFilters=catFilters)

[1] "8870 rows and 502 columns completed in : 12.761 seconds"


In [12]:
data

Patient.ID X.demographics.AGE. X.demographics.RACE.
1             5                  18              mexican
2            19                  12              mexican
3            22                  20              mexican
4            25                  18              mexican
5            26                  23              mexican
6            33                  12                white
7            38                  19              mexican
8            44                   5              mexican
9            45                  16              mexican
10           47                  17              mexican
11           48                  12                white
12           52                  11              mexican
13           53                  19                white
14           59                  19              mexican
15           61                  17                white
16           62                   8              mexican
17           66                  16                white
18           74                  24              mexican
19           85                  16                white
20           87                  11                white
21           88                  13                white
22           92                  18              mexican
23           93                  16              mexican
24           98                  22                white
25          102                  19              mexican
26          117                  18              mexican
27          123                  19              mexican
28          129                  19                white
29          130                  12                white
30          133                  19                white
⋮             ⋮                   ⋮                    ⋮
8841      41337                  17              mexican
8842      41338                  16              mexican
8843      41341                  17                white
8844      41344                  16              mexican
8845      41348                  13                white
8846      41351                  21              mexican
8847      41355                  12              mexican
8848      41375                  19              mexican
8849      41377                  15              mexican
8850      41378                  21                white
8851      41383                   9              mexican
8852      41386                  17              mexican
8853      41396                  13              mexican
8854      41398                   7                white
8855      41399                  13              mexican
8856      41401                  12              mexican
8857      41411                  22                white
8858      41413                  10              mexican
8859      41415                  12                white
8860      41417                  16              mexican
8861      41421                   6                white
8862      41422                  18              mexican
8863      41438                  24              mexican
8864      41439                   9              mexican
8865      41440                  19              mexican
8866      41444                  17              mexican
8867      41455                  15              mexican
8868      41462                  19              mexican
8869      41470                  18                white
8870      41474                   5                white
     X.examination.blood.pressure.60.sec.HR..30.sec.HR...2..
1                                                        NaN
2                                                        NaN
3                                                        NaN
4                                                        NaN
5                                                        NaN
6                                                        NaN
7                                           

# Now if we really just want all the data, we can do something like this

In [ ]:
data = runQueryAndGetResult(dictionaryFields(""))

In [ ]:
data